helix-core/src/date.rs

```diff
@@ -0,0 +1,422 @@
+}
+
+impl DateIncrementor {
+    pub fn from_range(text: RopeSlice, range: Range) -> Option<DateIncrementor> {
+        let range = if range.is_empty() {
+            if range.anchor < text.len_bytes() {
+                // Treat empty range as a cursor range.
+                range.put_cursor(text, range.anchor + 1, true)
+            } else {
+                // The range is empty and at the end of the text.
+                return None;
+            }
+        } else {
+            range
+        };
+
+        let from = range.from().saturating_sub(DATE_LENGTH);
+        let to = (range.from() + DATE_LENGTH).min(text.len_chars());
+
+        let (from_in_text, to_in_text) = (range.from() - from, range.to() - from);
+        let text: Cow<str> = text.slice(from..to).into();
+
+        FORMATS.iter().find_map(|&format| {
+            let re = Regex::new(format.regex).ok()?;
+            let captures = re.captures(&text)?;
+
+            let date = captures.get(0)?;
+            let offset = range.from() - from_in_text;
+            let range = Range::new(date.start() + offset, date.end() + offset);
+
+            let (year, month, day) = (captures.get(1)?, captures.get(2)?, captures.get(3)?);
+            let (year_range, month_range, day_range) = (year.range(), month.range(), day.range());
+
+            let field = if year_range.contains(&from_in_text)
+                && year_range.contains(&(to_in_text - 1))
+            {
+                DateField::Year
+            } else if month_range.contains(&from_in_text) && month_range.contains(&(to_in_text - 1))
+            {
+                DateField::Month
+            } else if day_range.contains(&from_in_text) && day_range.contains(&(to_in_text - 1)) {
+                DateField::Day
+            } else {
+                return None;
+            };
+
+            let date = Date::new(
+                year.as_str().parse::<i16>().ok()?,
+                month.as_str().parse::<u8>().ok()?,
+                day.as_str().parse::<u8>().ok()?,
+            )
+            .ok()?;
+
+            Some(DateIncrementor {
+                date,
+                field,
+                range,
+                format,
+            })
+        })
+    }
+
+    pub fn incremented_text(&self, amount: i64) -> Tendril {
+        let date = match self.field {
+            DateField::Year => self
+                .date
+                .add_years(amount.try_into().unwrap_or(0))
+                .or_next_valid(),
+            DateField::Month => self
+                .date
+                .add_months(amount.try_into().unwrap_or(0))
+                .or_prev_valid(),
+            DateField::Day => self.date.add_days(amount.try_into().unwrap_or(0)),
+        };
+
+        format!(
+            "{:04}{}{:02}{}{:02}",
+            date.year(),
+            self.format.separator,
+            date.month().to_number(),
+            self.format.separator,
+            date.day()
+        )
+        .into()
+    }
```

Comment:
Since we have both `DateIncrementor` and `NumberIncrementor` with the same function, can we separate this out into a trait? With this, we can probably also simplify the logic below to reduce duplication with something like `Box<dyn Incrementable>` and only put one `increment_text`.

Reply:
Yeah, we should definitely have a trait for this.